### PR TITLE
journal: fix _SOURCE_MONOTONIC_TIMESTAMP field

### LINKED ISSUE #33293: Kernel logs times are off by a bit more than one day
### systemd version the issue has been seen with

256-rc4

### Used distribution

Debian Trixie

### Linux kernel version used

6.7.12-amd64

### CPU architectures issue was seen on

None

### Component

systemd-journald

### Expected behaviour you didn't see

I expected "monotonic" times while reading logs from `journalctl -f` (at least with -o short-monotonic).

### Unexpected behaviour you saw

I saw logs from the future here, like from tomorrow, a tad more than 24h. From 24h 36m and a few seconds in the future, or 88563 seconds but I bet this number is not a constant.

### Steps to reproduce the problem

I don't know yet, I only have a few hints.

### Additional program output to the terminal or log subsystem illustrating the issue

Boot happened with: `Jun 10 18:33:27 kalir systemd-journald[1318]: Journal started`

First occurence of time travel a few hours later, related to hibernation:

```text
Jun 10 21:57:50 kalir NetworkManager[1742]: <info>  [1718049470.7180] dhcp4 (wlp1s0): state changed no lease
Jun 11 00:12:24 kalir kernel: Lockdown: systemd-logind: hibernation is restricted; see man kernel_lockdown.7
Jun 10 21:57:50 kalir dbus-daemon[1682]: [system] Activating via systemd: service name='org.freedesktop.nm_dispatcher' unit='dbus-org.freedesktop.nm-dispatcher.service' requested by ':1.14' (uid=0 pid=1742 comm="/usr/sbin/NetworkManager --no-daemon")
```

Now, at the time of writing, we're at `Jun 12 13:39:15`, but I already see events from the future by searching `Jun 13` in the logs:

```
Jun 12 13:29:15 kalir rtkit-daemon[2213]: Supervising 0 threads of 0 processes of 0 users.
Jun 12 13:29:15 kalir rtkit-daemon[2213]: Supervising 0 threads of 0 processes of 0 users.
Jun 12 13:29:31 kalir gdm-password][104683]: gkr-pam: unlocked login keyring
Jun 12 13:29:32 kalir NetworkManager[1742]: <info>  [1718191772.0770] agent-manager: agent[4117da6764403ca4,:1.78/org.gnome.Shell.NetworkAgent/1000]: a
gent registered
Jun 13 15:05:49 kalir kernel: [drm:mes_v11_0_submit_pkt_and_poll_completion.constprop.0 [amdgpu]] *ERROR* MES failed to response msg=14
Jun 13 15:05:49 kalir kernel: [drm:amdgpu_mes_reg_write_reg_wait [amdgpu]] *ERROR* failed to reg_write_reg_wait
Jun 13 15:05:49 kalir kernel: [drm:mes_v11_0_submit_pkt_and_poll_completion.constprop.0 [amdgpu]] *ERROR* MES failed to response msg=14
Jun 13 15:05:49 kalir kernel: [drm:amdgpu_mes_reg_write_reg_wait [amdgpu]] *ERROR* failed to reg_write_reg_wait
Jun 12 13:29:50 kalir rtkit-daemon[2213]: Supervising 0 threads of 0 processes of 0 users.
Jun 12 13:29:50 kalir rtkit-daemon[2213]: Supervising 0 threads of 0 processes of 0 users.
```

I notice that the `kernel` lines are in the future while the **not** kernel ones are good.

Running journalctl with `-o short-monotonic` does not help, I see the same jumps:

```
[ 4258.534832] kalir dbus-daemon[1682]: [system] Activating via systemd: service name='org.freedesktop.nm_dispatcher' unit='dbus-org.freedesktop.nm-dis
patcher.service' requested by ':1.14' (uid=0 pid=1742 comm="/usr/sbin/NetworkManager --no-daemon")
[ 4258.535896] kalir avahi-daemon[1679]: Withdrawing address record for 192.168.1.34 on wlp1s0.
[12332.553347] kalir kernel: Lockdown: systemd-logind: hibernation is restricted; see man kernel_lockdown.7
[ 4258.535939] kalir avahi-daemon[1679]: Leaving mDNS multicast group on interface wlp1s0.IPv4 with address 192.168.1.34.
[12332.557761] kalir kernel: Lockdown: systemd-logind: hibernation is restricted; see man kernel_lockdown.7
[12332.561601] kalir kernel: Lockdown: systemd-logind: hibernation is restricted; see man kernel_lockdown.7
[12332.564929] kalir kernel: Lockdown: systemd-logind: hibernation is restricted; see man kernel_lockdown.7
[ 4258.548293] kalir avahi-daemon[1679]: Interface wlp1s0.IPv4 no longer relevant for mDNS.
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #33293.

### CODE REVIEW

**Comment by yuwata** [comment]
The first commit needs to be backported to v256-stable.

**Review by bluca — APPROVED** [review]

**Comment by poettering** [comment]
guys, are we in a hurry? there is no need to hurry these things at all? you hurry this stuff into a stable release 12h after it was posted for review? cool down, please. If this was an important regression then maybe this was appropriate, but this is not important nor a regression, so why the hurry?

please, let's give non-trivial stuff like this a chance to be reviewed properly.

**Comment by bluca** [comment]
The ticket says it _is_ a regression (it's marked as such), and logs are being lost, which to me sounds kinda important and worth including in the stable release. And it was open the whole working day, and sitting with a green label for like 5 hours, so it's not like it all happened within 5 minutes...
Is there any issue with the approach taken?

**Comment by yuwata** [comment]
-> #33401.

File: src/shared/logs-show.c [hunk 1]
```diff
@@ -450,6 +450,9 @@ static void parse_display_realtime(
         assert(j);
         assert(ret);
 
+        // FIXME: _SOURCE_MONOTONIC_TIMESTAMP is in CLOCK_BOOTTIME, hence we cannot use it for adjusting realtime.
+        source_monotonic = NULL;
```
Comment (poettering) [inline src/shared/logs-show.c]: hmm, i am not sure i grok this comment. i mean we shouldn't really fix this up when parsing, but simply fill in the correct fields.

note there has been a long standing issue that we probably should include the CLOCK_BOOTTIME timestamp in all journal entries too (right now we only store the realtime + monotonic timestamp). There's even a bug open about that somewhere I think, was unable to find it quickly though.

I couldn't come up with a nice way to include the boottime timestamp in the records without massively breaking the format and increasing the entry record size so far. The least bad idea i had, is that depending on some feature flag we'd use the upper 32bit of seqnum, plus the upper 8bit of the monotonic + realtime fields of EntryObject to encode the offset of boottime towards seqnum. That would give us 32+8+8 = 48bit accuracy, i.e we'd allow 9 years or so of delta between CLOCK_MONOTONIC and CLOCK_BOOTTIME. i.e. people could hibernate their system 9 year long before the journal couldn't deal with it anymore.

**Comment by poettering** [comment]
maybe a "compatible" header field should indicate whether _SOURCE_MONOTONIC_TIMESTAMP actually means what it says on the tin, or is actually _SOURCE_BOOTTIME_TIMESTAMP. 

i.e. HEADER_COMPATIBLE_SOURCE_MONOTONIC_WORKS or so as new flag, next to the other HEADER_COMPATIBLE_* flags.

File: src/shared/logs-show.c (hunk 1 above)
Comment (yuwata) [reply src/shared/logs-show.c]: Yeah, I am considering something similar now. Saving shift is slightly complicated, but can keep entry object size, hence maybe good idea.

**Comment by yuwata** [comment]
> maybe a "compatible" header field should indicate whether _SOURCE_MONOTONIC_TIMESTAMP actually means what it says on the tin, or is actually _SOURCE_BOOTTIME_TIMESTAMP.
> 
> i.e. HEADER_COMPATIBLE_SOURCE_MONOTONIC_WORKS or so as new flag, next to the other HEADER_COMPATIBLE_* flags.

As commented in https://github.com/systemd/systemd/pull/33401#issuecomment-2178889963, I have no idea to make _SOURCE_MONOTONIC_TIMESTAMP reliable. If there is a way, then may be better to introduce a flag, of course.

**Pull request merged**